grub-core/fs/zfs/zfs.c

```diff
@@ -3765,6 +3757,60 @@ zfs_mtime (grub_device_t device, grub_int32_t *mt)
   return GRUB_ERR_NONE;
 }
 
+static grub_err_t
+zfs_devs_read_zbt (struct grub_zfs_data *data, grub_uint64_t offset, char *buf, grub_size_t len)
+{
+  grub_err_t err = GRUB_ERR_NONE;
+  zio_cksum_t zc;
+  unsigned int i;
+  ZIO_SET_CHECKSUM(&zc, offset, 0, 0, 0);
+
+  for (i = 0; i < data->n_devices_attached; i++)
+    {
+      err = grub_disk_read (data->devices_attached[i].dev->disk,
+			    offset >> SPA_MINBLOCKSHIFT,
+			    offset & ((1 << SPA_MINBLOCKSHIFT) - 1),
+			    len, buf);
+      if (err)
+	continue;
+      ZIO_SET_CHECKSUM(&zc, offset, 0, 0, 0);
+      err = zio_checksum_verify (zc, ZIO_CHECKSUM_LABEL, GRUB_ZFS_LITTLE_ENDIAN,
+				 buf, len);
+      if (!err)
+	return GRUB_ERR_NONE;
+    }
+  return err;
+}
+
+static grub_err_t
+grub_zfs_envblk_open (struct grub_file *file)
+{
+  grub_err_t err;
+  struct grub_zfs_data *data;
+  vdev_boot_envblock_t *vbe;
+  int l;
+
+  file->offset = 0;
+  data = zfs_mount (file->device);
+  file->data = data;
+  data->file_buf = grub_malloc (sizeof (vdev_boot_envblock_t));
+  for (l = 0; l < VDEV_LABELS / 2; l++)
+    {
+      grub_uint64_t offset = l * sizeof (vdev_label_t) + offsetof (vdev_label_t, vl_be);
+
+      err = zfs_devs_read_zbt (data, offset, data->file_buf,
+			       sizeof (vdev_boot_envblock_t));
+      if (err == GRUB_ERR_NONE)
+	{
+	  vbe = (vdev_boot_envblock_t *)data->file_buf;
+	  file->size = grub_strlen (vbe->vbe_bootenv);
```

Comment:
Since we write them out from first to last and check them from first to last, the first one that is valid should be the latest one. If a write gets dropped by the disk it's possible that this could be incorrect, I guess. We could put a generation number or TXG in there, if we think that's a significant issue.